contracts/rewards/src/contract/execute.rs

```diff
@@ -77,113 +77,106 @@ where
+                Ok(LoadState::New(event))
+            }
+            Some(event) => Ok(LoadState::Loaded(event)),
         }
-        .record_participation(worker);
-
-        self.store.save_epoch_tally(&tally)?;
-
-        Ok(())
     }
 
     pub fn distribute_rewards(
         &mut self,
-        contract: Addr,
+        contract_addr: Addr,
         cur_block_height: u64,
         epoch_process_limit: Option<u64>,
     ) -> Result<HashMap<Addr, Uint128>, ContractError> {
         let epoch_process_limit = epoch_process_limit.unwrap_or(DEFAULT_EPOCHS_TO_PROCESS);
         let cur_epoch = self.current_epoch(cur_block_height)?;
+
         let from = self
             .store
-            .load_rewards_watermark(contract.clone())?
+            .load_rewards_watermark(contract_addr.clone())?
             .map_or(0, |last_processed| last_processed + 1);
+
         let to = std::cmp::min(
-            (from + epoch_process_limit).saturating_sub(1),
-            cur_epoch.epoch_num.saturating_sub(2),
+            (from + epoch_process_limit).saturating_sub(1), // for process limit =1 "from" and "to" must be equal
+            cur_epoch.epoch_num.saturating_sub(EPOCH_PAYOUT_DELAY),
         );
-        if to < from || cur_epoch.epoch_num < 2 {
+
+        if to < from || cur_epoch.epoch_num < EPOCH_PAYOUT_DELAY {
             return Err(ContractError::NoRewardsToDistribute.into());
         }
-        let rewards =
-            self.process_rewards_for_epochs(contract.clone(), from, to, HashMap::new())?;
-        self.store.save_rewards_watermark(contract, to)?;
+
+        let rewards = self.process_rewards_for_epochs(contract_addr.clone(), from, to)?;
+        self.store.save_rewards_watermark(contract_addr, to)?;
         Ok(rewards)
     }
 
     fn process_rewards_for_epochs(
         &mut self,
-        contract: Addr,
+        contract_addr: Addr,
         from: u64,
         to: u64,
-        rewards: HashMap<Addr, Uint128>,
-    ) -> Result<HashMap<Addr, Uint128>, ContractError> {
-        if from > to {
-            return Ok(rewards);
-        }
-        let new_rewards = self.process_rewards_for_epoch(contract.clone(), from)?;
-        self.process_rewards_for_epochs(contract, from + 1, to, merge_rewards(rewards, new_rewards))
-    }
-
-    pub fn process_rewards_for_epoch(
-        &mut self,
-        contract: Addr,
-        epoch_num: u64,
     ) -> Result<HashMap<Addr, Uint128>, ContractError> {
         self.store
-            .load_epoch_tally(contract, epoch_num)?
-            .map_or(Ok(HashMap::new()), |tally| self.process_epoch_tally(tally))
+            .load_rewards_pool(contract_addr.clone())?
+            .sub_reward(
+                self.cumulate_rewards(&contract_addr, from, to)
+                    .values()
+                    .sum(),
+            )?
+            .then(|pool| self.store.save_rewards_pool(&pool))?;
+
+        Ok(self.cumulate_rewards(&contract_addr, from, to))
```

Comment:
Am I misreading this, or are we calling `self.cumulate_rewards` twice with the exact same parameters? If so, I don't think we should do that, for gas and performance reasons. Can we just cache the cumulated value somehow?

Reply:
oh, line 162 just shouldn't be there!